### ISSUE TITLE
DUDAS I1

### ISSUE THREAD

**Issue by tatanramos** [IIC2343/Syllabus-2021-2-S1#10]
Hola! Les dejo las dudas de la i1 que fueron saliendo durante la sesión de zoom.
### **Dudas generales**

- La pregunta 0 la pueden adjuntar en cualquier pregunta, pero tiene que estar
- Las preguntas se suben a cuestionarios diferentes
- Tienen que contestar solo 3 preguntas
- Toda la prueba se responde escrita (manuscrita, usando el computador, etc.), que esté lo más ordenado posible.
- en formato pdf

### **Pregunta 1**
(Duda general)
_"A qué se refiere con el multiplexor del computador de valeria? deberíamos estar pensando en algún diseño especifico de computador?"_

- Son nombres de fantasía nomás.

### **1a)** 

_"Es el mismo bit erróneo para todos?"_

-  No necesariamente es el mismo bit erróneo.

_"Es necesario mostrar como llegamos al resultado?"_

- Sí, pero deduciendolo de forma "simple" (No es necesaria una argumentación matemática)

_"El número de la población de votantes está en base 10?"_

- Sí

_"Cuando hablan de bit erróneo se refiere a que siempre está equivocado? No significa que , por ejemplo, el bit esté “pegado” en un valor (0 o 1) y a veces coincide y otras no"_

- Es un bit que siempre cambia el valor que debería tener.

_"El bit malo es en el resultado o puede estar en algún otro lado como en algún otro lado como en algún bit para la control unit que afectó el conteo per sé?"_

- Sí, está en el resultado

_"No deberíamos preocuparnos por la arquitectura interna, solo del resultado cierto?"_

- Efectivamente, solo del resultado

### **1b)**
(Comentario general)
Pueden usar los circuitos que usamos en clases. En caso de que usen algo especial expliquenlo un poco

_"Podría hacer un circuito que se base en cada bit?"_

- pueden hacer el circuito que estimen conveniente

_"El circuito de la 1b es como un computador básico? o con compuertas lógicas?"_

- Es con compuertas lógicas 

_"El circuito debe funcionar para este caso particular de votaciones, o se deben tomar en cuenta, por ejemplo, alguna futura votación con los mismos computadores y mismos errores,  pero otros números? "_

- Se corrige para la situación particular del ejercicio.

_"Cuando dicen un solo circuito aplicable a ambos computadores, se refiere a un circuito que reciba ambos inputs (diferenciables) y los corrija?"_

- Tienes que poder poner el mismo circuito a la salida de los 2 computadores, y en ambos casos debe ser capaz de corregir el error específico

### **1c)**
(Comentario general)
Esta pregunta es independiente de la 1a y 1b

### **Pregunta 2**
(Duda general)
_"Hay alguna cantidad máxima de inputs en las compuertas (and, or)"_

- Pueden usar el número que estimen conveniente de inputs para esas compuertas.

### **2a)** 
_"Cuando nos dicen que queremos “tener acceso a la Data Memory a través de una dirección base más un desplazamiento.”, se refiere a que queremos acceder a ella como si fuese un arreglo?"_

- Sí, esa es la idea

_" Se debe diseñar el circuito entregando tanto r como e o solo r?"_

- Debe entregar r y e

_"En la 2a cuando e = 1, r puede tomar valor de 1 o 0 ?"_

- Exactamente, r puede tomar cualquiera de esos valores

_"Cuando hay empate da lo mismo que entrega el r?"_

- Da lo mismo

_"El valor que toma r en empates tiene que ser consistente o en algunos casos puede ser 1 y en otros 0?"_

- Es irrelevante, lo que importa es que hay empate, entonces el "e" está en 1.

_"El output x del cable r, se refiere x a que x puede ser tanto 1 como 0? x es que da lo mismo que bit es?"_

- Ese x es porque da lo mismo si es un 0 o un 1.

### **2b)**
_"Tenemos que reescribir el circuito en a? "_

- Pueden poner una "caja" que simbolice que ese circuito está creado, aunque no lo hayan dibujado en la parte a) correctamente

### **Pregunta 3**
_"A qué se refiere con que el registro D es especial?, en que se diferencia de los registros A y B ?"_

- Es especial porque solo guarda valores de desplazamientos

_"Entonces el D no va a entrar a la ALU?"_

- Por ahora no entra a la ALU, puede ser escrito, pero no es leído

_"El registro D almacena un valor que le llega por instruction Memory?"_

- Sí

_"el desplazamiento es positivo ? o también puede ser negativo ?"_

- Es positivo.

### **Pregunta 4**
_"Cuales son todas las instrucciones que podemos usar para la 4?"_

- Se usa el assembly del computador básico **(no riscv)**

_"En la pregunta 4 p y q son únicos inputs? o tenemos r también?"_

- Sí, pero en la 4b hay un input extra.  el programa debe determinar r

_"Asumimos que el p y el q están en los registros?"_

- p y q pueden estar en la data memory, los tienen que pasar a los registros cuando hagan su código.

**Comment by tatanramos** [comment]
Les dejo al bueno de Ben para que les vaya excelente!!
![brereton](https://user-images.githubusercontent.com/69760120/135094128-4ab1d6e6-eab8-47c4-8336-6c54324925a7.jpeg)

**Comment by Mario-uc** [comment]
una consulta, en la 2b como son muchas conexiones, ¿podemos dibujar el circuito como lo hacia el profe cuando nos pasó la materia de procesador? (es decir, no dibujar directamente las conexiones sino escribir que inputs entran en cada compuerta)

**Comment by asuribe** [comment]
en la pregunta 2b), puedo usar los valores de r y e como inputs del circuito para r5??

**Comment by mmsaavedra1** [comment]
Hola, en la 4 debo retornar m para cualquiera de los casos (q== 2^r) o (q != 2^r)?

**Comment by fernando98alba** [comment]
Hola! en la 4) puedo asumir que hay una instruccion para multiplicar?

**Comment by sufigueroa** [comment]
> una consulta, en la 2b como son muchas conexiones, ¿podemos dibujar el circuito como lo hacia el profe cuando nos pasó la materia de procesador? (es decir, no dibujar directamente las conexiones sino escribir que inputs entran en cada compuerta)

Siii solamente debe quedar claro que hace el circuito

**Comment by sufigueroa** [comment]
> Hola, en la pregunta 2 podemos usar la tecnica de Mapas de Karnaugh para reducir los términos encontrados?

Sip, pero no importa mucho que quede eficiente por lo que no recomendaria hacer Karnaugh para el ejercicio :)

**Comment by sufigueroa** [comment]
> en la pregunta 2b), puedo usar los valores de r y e como inputs del circuito para r5??

Siii recomiendo hacerlo

**Comment by sufigueroa** [comment]
> Hola, en la 4 debo retornar m para cualquiera de los casos (q== 2^r) o (q != 2^r)?

Sii la idea es hacer una multiplicacion inteligente por lo que debe servir para q != 2^r tambien

**Comment by sufigueroa** [comment]
> Hola! en la 4) puedo asumir que hay una instruccion para multiplicar?

No, tu debes hacer la multiplicacion utilizando las instrucciones del computador basico visto en clases (no RISC-V). Para la parte de calcular el factorial puedes asumir que la funcion de la primera parte funciona

**Comment by MGilB** [comment]
Para la pregunta 4 puedo asumir que las variables p, q, r, s son registros o tenemos que hacer un programa que vaya a buscar a la memoria esos valores y solo trabaje con registros A y B

**Comment by JuaniAndres** [comment]
Hola para la pregunta 4 podemos definir más variables? y es necesario dejar guardado el valor de r o da lo mismo si llegamos igualmente a calcular m?

**Comment by itplanella** [comment]
Hola, en la pregunta 4 solo podemos usar 2 registros?

**Comment by sufigueroa** [comment]
> Para la pregunta 4 puedo asumir que las variables p, q, r, s son registros o tenemos que hacer un programa que vaya a buscar a la memoria esos valores y solo trabaje con registros A y B

Solo tienes los registros A y B en el computador basico, p y q pueden estar en memoria

**Comment by sufigueroa** [comment]
> Hola para la pregunta 4 podemos definir más variables? y es necesario dejar guardado el valor de r o da lo mismo si llegamos igualmente a calcular m?

Puedes usar todas las variables que estimes necesario y no es necesario dejar guardado r

**Comment by sufigueroa** [comment]
> Hola, en la pregunta 4 solo podemos usar 2 registros?

Sip, el computador basico solo tiene A y B

**Comment by JuaniAndres** [comment]
En la pregunta 4, p y q pueden ser negativos o 0?

**Comment by MarceloSaldias** [comment]
Hola! En la 2a) se me hace un poco dificil dibujar todo el detalle del circuito en un solo dibujo, puedo dibujar en otro lado (y con mas espacio) algunos de los circuitos y después en otro dibujo mostrar el circuito general? (Como lo hacía el profe en clases con los half-adder y después full-adder, por ejemplo)

**Comment by RodrigoNazar** [comment]
En la pregunta 3, en la parte b necesitamos escribir los nuevos opcodes que se utilizaran, ¿pueden ser más de una instrucción adicional? ¿hay alguna restricción de la cantidad de ciclos de clock que podemos usar para realizar la operación que se pide? ¿Se espera que creemos una instrucción del estilo ACC B, D? lo que significaría acceder a la dirección que está en B desplazada D veces y cargar el valor al registro B.

**Comment by crvargasu** [comment]
Se puede usar https://logic.ly/demo/ para diagramar?

**Comment by Fernan-tess** [comment]
Hola, En la 3 se supone que se suman los registros de B y D, y el resultado se ingresa, según corresponda, a la data memory?

**Comment by camaruri** [comment]
Hola, la misma pregunta que @MarceloSaldias, podríamos dibujar varios circuitos y después usarlos en el circuito general o usar el output de cada circuito en otro de los circuitos apartes?

**Comment by tomascortes** [comment]
Hola!
En la 4a
tenemos que considerar el caso en que se multiplique por 0?

**Comment by pedroherreraj** [comment]
En la 3.b, es necesario implementar la operacion ADD D, (cualquier valor)? Por que el registro D no pasa por la ALU, por lo que se complicaría un poco la respuesta

**Comment by acastill20** [comment]
En la 4.b el unico input es s y de ahi tenemos que ir ideando el p y el q o p y q son entregados previamente?

**Comment by Vinuelax** [comment]
> En la pregunta 4, p y q pueden ser negativos o 0? Me entra la duda ya que dice números enteros pero en todos los ejemplos ocupan números positivos

pueden asumir que seran positivos

**Comment by ferperezh** [comment]
hola en la 1a) el bit del computador de carlos se puede cambiar un bit n q no necesariamente es 1 por el 0?, es que el enunciado solo dice de 1 a 0 o de 0 a 1

**Comment by Vinuelax** [comment]
> > En la pregunta 4, p y q pueden ser negativos o 0? Me entra la duda ya que dice números enteros pero en todos los ejemplos ocupan números positivos
> 
> pueden asumir que seran positivos

Me corrijo pueden ser positivos o 0

**Comment by Vinuelax** [comment]
> Hola! En la 2a) se me hace un poco dificil dibujar todo el detalle del circuito en un solo dibujo, puedo dibujar en otro lado (y con mas espacio) algunos de los circuitos y después en otro dibujo mostrar el circuito general? (Como lo hacía el profe en clases con los half-adder y después full-adder, por ejemplo)

Mientras se entiendan y no sea demasiado complejo no deberia haber problemas

**Comment by Vinuelax** [comment]
> En la pregunta 3, en la parte b necesitamos escribir los nuevos opcodes que se utilizaran, ¿pueden ser más de una instrucción adicional? ¿hay alguna restricción de la cantidad de ciclos de clock que podemos usar para realizar la operación que se pide? ¿Se espera que creemos una instrucción del estilo ACC B, D? lo que significaría acceder a la dirección que está en B desplazada D veces y cargar el valor al registro B.

La idea es que crees todas las instrucciones necesarias para hacer lo que te pide el enunciado, no hay un minimo ni maximo

**Comment by Vinuelax** [comment]
> Hola, En la 3 se supone que se suman los registros de B y D, y el resultado se ingresa, según corresponda, a la data memory?

Se espera que pueda ser utilizada para poder utilizar la direccion generada

**Comment by Vinuelax** [comment]
> En la 3.b, es necesario implementar la operacion ADD D, (cualquier valor)? Por que el registro D no pasa por la ALU, por lo que se complicaría un poco la respuesta

No, no es necesario. La idea es que sea un registro especializado para esa funcion.

**Comment by Vinuelax** [comment]
> En la 4.b el unico input es s y de ahi tenemos que ir ideando el p y el q o p y q son entregados previamente?

El unico input es s, y al momento de querer calcular el factorial de s deberias porder saber cuales son los valores de p y q para la funcion mult
